### PR TITLE
feat: add Pebble DB backend

### PR DESCRIPTION
**Pull request by wolkykim**
- add Pebble DB backend for experiments
- related to #1013

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

Benchmark output, based on the random read/write test patterns we have, shows faster op time compared to the default LevelDB backend. Note that this is Async read/write)
```
$ go test -cpu 1,4,10 -benchmem -run=^$ -bench=. -tags pebbledb
goos: darwin
goarch: arm64
pkg: github.com/gnolang/gno/tm2/pkg/db
BenchmarkGoLevelDBRandomReadsWrites       	  313208	      4180 ns/op	     672 B/op	      13 allocs/op
BenchmarkGoLevelDBRandomReadsWrites-4     	  307611	      4138 ns/op	     680 B/op	      13 allocs/op
BenchmarkGoLevelDBRandomReadsWrites-10    	  316270	      4081 ns/op	     679 B/op	      13 allocs/op

BenchmarkPebbleDBRandomReadsWrites          	  359060	      3742 ns/op	      41 B/op	       3 allocs/op
BenchmarkPebbleDBRandomReadsWrites-4        	  376674	      3335 ns/op	      32 B/op	       3 allocs/op
BenchmarkPebbleDBRandomReadsWrites-10       	  390951	      3384 ns/op	      32 B/op	       3 allocs/op

## When PebbleDB is tuned to use 16MB cache (default 8MB)
BenchmarkPebbleDBRandomReadsWrites          	  347396	      3470 ns/op	      42 B/op	       3 allocs/op
BenchmarkPebbleDBRandomReadsWrites-4        	  368671	      3011 ns/op	      31 B/op	       3 allocs/op
BenchmarkPebbleDBRandomReadsWrites-10       	  346952	      3000 ns/op	      32 B/op	       3 allocs/op
```

```
$ go test . -tags pebbledb
ok  	github.com/gnolang/gno/tm2/pkg/db	2.151s

$ go test . -tags pebbledb -v | grep -i pebbledb
=== RUN   TestBackendsGetSetDelete/pebbledb
    --- PASS: TestBackendsGetSetDelete/pebbledb (0.06s)
=== RUN   TestBackendsNilKeys/Testing_pebbledb
    --- PASS: TestBackendsNilKeys/Testing_pebbledb (0.08s)
=== RUN   TestDBIterator/pebbledb
    --- PASS: TestDBIterator/pebbledb (0.03s)
=== RUN   TestDBIteratorSingleKey/Backend_pebbledb
    --- PASS: TestDBIteratorSingleKey/Backend_pebbledb (0.04s)
=== RUN   TestDBIteratorTwoKeys/Backend_pebbledb
    --- PASS: TestDBIteratorTwoKeys/Backend_pebbledb (0.05s)
=== RUN   TestDBIteratorMany/Backend_pebbledb
    --- PASS: TestDBIteratorMany/Backend_pebbledb (0.03s)
=== RUN   TestDBIteratorEmpty/Backend_pebbledb
    --- PASS: TestDBIteratorEmpty/Backend_pebbledb (0.04s)
=== RUN   TestDBIteratorEmptyBeginAfter/Backend_pebbledb
    --- PASS: TestDBIteratorEmptyBeginAfter/Backend_pebbledb (0.04s)
=== RUN   TestDBIteratorNonemptyBeginAfter/Backend_pebbledb
    --- PASS: TestDBIteratorNonemptyBeginAfter/Backend_pebbledb (0.04s)
=== RUN   TestPebbleDB
--- PASS: TestPebbleDB (0.04s)
=== RUN   TestPrefixIteratorNoMatchNil/Prefix_w/_backend_pebbledb
    --- PASS: TestPrefixIteratorNoMatchNil/Prefix_w/_backend_pebbledb (0.04s)
=== RUN   TestPrefixIteratorNoMatch1/Prefix_w/_backend_pebbledb
    --- PASS: TestPrefixIteratorNoMatch1/Prefix_w/_backend_pebbledb (0.05s)
=== RUN   TestPrefixIteratorNoMatch2/Prefix_w/_backend_pebbledb
    --- PASS: TestPrefixIteratorNoMatch2/Prefix_w/_backend_pebbledb (0.04s)
=== RUN   TestPrefixIteratorMatch1/Prefix_w/_backend_pebbledb
    --- PASS: TestPrefixIteratorMatch1/Prefix_w/_backend_pebbledb (0.05s)
=== RUN   TestPrefixIteratorMatches1N/Prefix_w/_backend_pebbledb
    --- PASS: TestPrefixIteratorMatches1N/Prefix_w/_backend_pebbledb (0.06s)
```

### CODE REVIEW

**Review by moul — APPROVED** [review]

**Comment by moul** [comment]
LGTM, could you rebase master and go mod tidy, please?

File: tm2/pkg/db/pebble_db.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+// Implements DB.
+func (db *PebbleDB) Get(key []byte) []byte {
+	value, closer, err := db.db.Get(nonNilBytes(key))
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return nil
+		}
+		panic(err)
+	}
```
Comment (ajnavarro) [inline tm2/pkg/db/pebble_db.go:62]: Avoid unnecessary nesting:

```suggestion
	if errors.Is(err, pebble.ErrNotFound) {
		return nil
	}
	if err != nil {
		panic(err)
	}
```

File: tm2/pkg/db/pebble_db.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+// Implements DB.
+func (db *PebbleDB) Has(key []byte) bool {
+	_, closer, err := db.db.Get(nonNilBytes(key))
+	if err != nil {
+		if errors.Is(err, pebble.ErrNotFound) {
+			return false
+		}
+		panic(err)
+	}
```
Comment (ajnavarro) [inline tm2/pkg/db/pebble_db.go:75]: Same as previous comment here

File: tm2/pkg/db/pebble_db.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+// Implements Batch.
+func (ba *pebbleDBBatch) Set(key, value []byte) {
+	ba.batch.Set(nonNilBytes(key), nonNilBytes(value), nil)
+}
+
+// Implements Batch.
+func (ba *pebbleDBBatch) Delete(key []byte) {
+	ba.batch.Delete(nonNilBytes(key), nil)
```
Comment (ajnavarro) [inline tm2/pkg/db/pebble_db.go:156]: unhandled error.

File: tm2/pkg/db/pebble_db.go (hunk 1 above)
Comment (wolkykim) [reply tm2/pkg/db/pebble_db.go:62]: I think for most of cases where err has nil value, the suggested logic will cost more as it goes through both L57 and L60.

File: tm2/pkg/db/pebble_db.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+type pebbleDBBatch struct {
+	db    *PebbleDB
+	batch *pebble.Batch
+}
+
+// Implements Batch.
+func (ba *pebbleDBBatch) Set(key, value []byte) {
+	ba.batch.Set(nonNilBytes(key), nonNilBytes(value), nil)
```
[the remark this answers is not in the document]
Comment (wolkykim) [reply tm2/pkg/db/pebble_db.go]: I've added an error check on this.

File: tm2/pkg/db/pebble_db.go (hunk 3 above)
Comment (wolkykim) [reply tm2/pkg/db/pebble_db.go]: also addressed

**Comment by wolkykim** [comment]
@moul just so you know, I see you updated the go.mod file and now it builds against the latest of the pebbled's master branch. Did you intend to do so? Previously I pinned the version to use the `latest crl-release-23.1` version. Anyway, there are small breaking changes in their iterator's return and I've updated the code accordingly. If you didn't mean it we can pin it down to isolate their change to affect gno build. FYI.

File: tm2/pkg/db/pebble_db.go (hunk 1 above)
Comment (ajnavarro) [reply tm2/pkg/db/pebble_db.go:62]: > I honestly don't see why the suggestion is a better way.

It is easier to read. If the code path is so hot that an extra `if` check will affect performance meaningfully, should we add some benchmarks to be careful about that?

Comment (moul) [reply tm2/pkg/db/pebble_db.go:62]: Certainly, prioritizing readability and simplicity is beneficial, even in smaller cases. By consistently emphasizing clarity, we foster good habits and ensure our work remains accessible to others. It's a commendable practice to always strike a balance between efficiency and readability.

See https://github.com/gnolang/gno/pull/1111

**Pull request closed without merging**